source/distributed/tria.cc

```diff
@@ -1124,10 +1124,545 @@ namespace
 } // namespace
 
 
+
 namespace parallel
 {
   namespace distributed
   {
+    /* ------------------ class DataTransfer<dim,spacedim> ----------------- */
+
+
+    template <int dim, int spacedim>
+    Triangulation<dim, spacedim>::DataTransfer::DataTransfer(
+      MPI_Comm mpi_communicator)
+      : mpi_communicator(mpi_communicator)
+    {}
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::DataTransfer::pack_data(
+      const std::vector<quadrant_cell_relation_t> &quad_cell_relations,
+      const std::vector<typename CellAttachedData::pack_callback_t>
+        &pack_callbacks)
+    {
+      Assert(src_data_fix.capacity() == 0,
+             ExcMessage("Previously packed data has not been released yet!"));
+
+      // Prepare the buffer structure, in which each callback function will
+      // store its data for each active cell.
+      // The outmost shell in this container construct corresponds to the
+      // data packed per cell. The next layer resembles the data that
+      // each callback function packs on the corresponding cell. These
+      // buffers are chains of chars stored in an std::vector<char>.
+      // A visualisation of the data structure:
+      /* clang-format off */
+      // |             cell_1                | |             cell_2                | ...
+      // ||  callback_1  ||  callback_2  |...| ||  callback_1  ||  callback_2  |...| ...
+      // |||char|char|...|||char|char|...|...| |||char|char|...|||char|char|...|...| ...
+      /* clang-format on */
+      std::vector<std::vector<std::vector<char>>> packed_data(
+        quad_cell_relations.size());
+
+      // Iterate over all cells, call all callback functions on each cell,
+      // and store their data in the corresponding buffer scope.
+      {
+        auto cit_qcrel    = quad_cell_relations.cbegin();
+        auto it_data_cell = packed_data.begin();
+        for (; cit_qcrel != quad_cell_relations.cend();
+             ++cit_qcrel, ++it_data_cell)
+          {
+            const auto &cell_status = std::get<1>(*cit_qcrel);
+            const auto &dealii_cell = std::get<2>(*cit_qcrel);
+
+#  ifdef DEBUG
```

Comment:
All of the things you have in between the preprocessor block will be optimized out in release mode. Just remove the `#ifdef ... #endif` directives.